Test failure: ManualRestartTest.when_terminalSnapshotFails_then_previousSnapshotUsed

https://hazelcast-l337.ci.cloudbees.com/job/jet-pr-builder/com.hazelcast.jet$hazelcast-jet-core/3082/testReport/junit/com.hazelcast.jet.core/ManualRestartTest/when_terminalSnapshotFails_then_previousSnapshotUsed/

Simplified error message:

```
{0=2, ... 2337=2, 2338=1, .. 9999=1} expected:<1> but was:<2>

Stacktrace AssertionError: expected:<1> but was:<2>
    at org.junit.Assert.fail(Assert.java:88)
    at org.junit.Assert.failNotEquals(Assert.java:834)
    at org.junit.Assert.assertEquals(Assert.java:118)
    at com.hazelcast.jet.core.ManualRestartTest.when_terminalSnapshotFails_then_previousSnapshotUsed(ManualRestartTest.java:188)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.base/java.lang.reflect.Method.invoke(Method.java:564)
    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
    at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
    at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:105)
    at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:97)
    at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.base/java.lang.Thread.run(Thread.java:844)```

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/MasterContext.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/MasterContext.java
@@ -505,7 +505,8 @@ public class MasterContext {
 
     private void cancelExecutionInvocations(long jobId, long executionId, TerminationMode mode) {
         nodeEngine.getExecutionService().execute(ExecutionService.ASYNC_EXECUTOR, () ->
-                invokeOnParticipants(plan -> new TerminateExecutionOperation(jobId, executionId, mode), responses -> { }, null));
+                invokeOnParticipants(plan -> new TerminateExecutionOperation(jobId, executionId, mode),
+                        responses -> { }, null));
     }
 
     void beginSnapshot(long executionId) {
@@ -541,7 +542,8 @@ public class MasterContext {
         Function<ExecutionPlan, Operation> factory =
                 plan -> new SnapshotOperation(jobId, executionId, newSnapshotId, isTerminal);
 
-        invokeOnParticipants(factory, responses -> onSnapshotCompleted(responses, executionId, newSnapshotId, isTerminal), null);
+        invokeOnParticipants(factory,
+                responses -> onSnapshotCompleted(responses, executionId, newSnapshotId, isTerminal), null);
     }
 
     private void onSnapshotCompleted(Map<MemberInfo, Object> responses, long executionId, long snapshotId,

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/MasterContext.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/MasterContext.java
@@ -828,21 +828,7 @@ public class MasterContext {
         for (MemberInfo member : executionPlanMap.keySet()) {
             responses.put(member, new AtomicReference<>());
         }
-
-        // we'll expect two callbacks for each member
-        AtomicInteger remainingCount = new AtomicInteger(2 * executionPlanMap.size());
-
-        Runnable handleDone = () -> {
-            if (remainingCount.decrementAndGet() == 0) {
-                Map<MemberInfo, Object> responses2 = new HashMap<>();
-                // unwrap the AtomicReferences. Can't use stream api because toMap doesn't support null values...
-                for (Entry<MemberInfo, AtomicReference<Object>> entry : responses.entrySet()) {
-                    responses2.put(entry.getKey(), entry.getValue().get());
-                }
-                completionCallback.accept(responses2);
-            }
-        };
-
+        AtomicInteger remainingCount = new AtomicInteger(executionPlanMap.size());
         for (Entry<MemberInfo, AtomicReference<Object>> e : responses.entrySet()) {
             MemberInfo member = e.getKey();
             AtomicReference<Object> response = e.getValue();
@@ -850,11 +836,18 @@ public class MasterContext {
             InternalCompletableFuture<Object> future =
                     nodeEngine.getOperationService()
                               .createInvocationBuilder(JetService.SERVICE_NAME, op, member.getAddress())
-                              .setDoneCallback(handleDone)
                               .setExecutionCallback(callbackOf((r, throwable) -> {
                                   boolean success = response.compareAndSet(null, r == null ? throwable : r);
                                   assert success : "response=" + response.get();
-                                  handleDone.run();
+                                  if (remainingCount.decrementAndGet() == 0) {
+                                      // unwrap the AtomicReferences. Can't use stream api because toMap doesn't support null values...
+                                      Map<MemberInfo, Object> responses2 = new HashMap<>();
+                                      for (Entry<MemberInfo, AtomicReference<Object>> entry : responses.entrySet()) {
+                                          responses2.put(entry.getKey(), entry.getValue().get());
+                                      }
+                                      completionCallback.accept(responses2);
+                                  }
+
                               }))
                               .invoke();
             if (callback != null) {

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/MasterContext.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/MasterContext.java
@@ -104,15 +104,15 @@ import static java.util.stream.Collectors.toList;
  */
 public class MasterContext {
 
+    public static final int SNAPSHOT_RESTORE_EDGE_PRIORITY = Integer.MIN_VALUE;
+    public static final String SNAPSHOT_VERTEX_PREFIX = "__snapshot_";
+
     private static final Object NULL_OBJECT = new Object() {
         @Override
         public String toString() {
             return "NULL_OBJECT";
         }
     };
-
-    public static final int SNAPSHOT_RESTORE_EDGE_PRIORITY = Integer.MIN_VALUE;
-    public static final String SNAPSHOT_VERTEX_PREFIX = "__snapshot_";
 
     private final Object lock = new Object();
 
@@ -513,8 +513,7 @@ public class MasterContext {
 
     private void cancelExecutionInvocations(long jobId, long executionId, TerminationMode mode) {
         nodeEngine.getExecutionService().execute(ExecutionService.ASYNC_EXECUTOR, () ->
-                invokeOnParticipants(plan -> new TerminateExecutionOperation(jobId, executionId, mode),
-                        null, null));
+                invokeOnParticipants(plan -> new TerminateExecutionOperation(jobId, executionId, mode), null, null));
     }
 
     void beginSnapshot(long executionId) {

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/MasterContext.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/MasterContext.java
@@ -840,7 +840,7 @@ public class MasterContext {
                                   boolean success = response.compareAndSet(null, r == null ? throwable : r);
                                   assert success : "response=" + response.get();
                                   if (remainingCount.decrementAndGet() == 0) {
-                                      // unwrap the AtomicReferences. Can't use stream api because toMap doesn't support null values...
+                                      // unwrap the AtomicReferences. Can't use stream api, toMap doesn't allow null values
                                       Map<MemberInfo, Object> responses2 = new HashMap<>();
                                       for (Entry<MemberInfo, AtomicReference<Object>> entry : responses.entrySet()) {
                                           responses2.put(entry.getKey(), entry.getValue().get());

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/MasterContext.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/MasterContext.java
@@ -62,9 +62,9 @@ import java.util.Map.Entry;
 import java.util.Set;
 import java.util.concurrent.CancellationException;
 import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Consumer;
 import java.util.function.Function;
 import java.util.stream.Collectors;
@@ -89,8 +89,8 @@ import static com.hazelcast.jet.impl.execution.SnapshotContext.NO_SNAPSHOT;
 import static com.hazelcast.jet.impl.execution.init.CustomClassLoadedObject.deserializeWithCustomClassLoader;
 import static com.hazelcast.jet.impl.execution.init.ExecutionPlanBuilder.createExecutionPlans;
 import static com.hazelcast.jet.impl.util.ExceptionUtil.isRestartableException;
-import static com.hazelcast.jet.impl.util.ExceptionUtil.peel;
 import static com.hazelcast.jet.impl.util.ExceptionUtil.withTryCatch;
+import static com.hazelcast.jet.impl.util.Util.callbackOf;
 import static com.hazelcast.jet.impl.util.Util.jobNameAndExecutionId;
 import static java.util.Collections.emptyList;
 import static java.util.concurrent.TimeUnit.NANOSECONDS;
@@ -352,7 +352,7 @@ public class MasterContext {
         Function<ExecutionPlan, Operation> operationCtor = plan ->
                 new InitExecutionOperation(jobId, executionId, membersView.getVersion(), participants,
                         nodeEngine.getSerializationService().toData(plan));
-        invoke(operationCtor, this::onInitStepCompleted, null);
+        invokeOnParticipants(operationCtor, this::onInitStepCompleted, null);
     }
 
     private void rewriteDagWithSnapshotRestore(DAG dag, long snapshotId) {
@@ -483,7 +483,7 @@ public class MasterContext {
 
         jobStatus = RUNNING;
 
-        invoke(operationCtor, completionCallback, executionInvocationCallback);
+        invokeOnParticipants(operationCtor, completionCallback, executionInvocationCallback);
 
         if (isSnapshottingEnabled()) {
             coordinationService.scheduleSnapshot(jobId, executionId);
@@ -505,7 +505,7 @@ public class MasterContext {
 
     private void cancelExecutionInvocations(long jobId, long executionId, TerminationMode mode) {
         nodeEngine.getExecutionService().execute(ExecutionService.ASYNC_EXECUTOR, () ->
-                invoke(plan -> new TerminateExecutionOperation(jobId, executionId, mode), responses -> { }, null));
+                invokeOnParticipants(plan -> new TerminateExecutionOperation(jobId, executionId, mode), responses -> { }, null));
     }
 
     void beginSnapshot(long executionId) {
@@ -541,7 +541,7 @@ public class MasterContext {
         Function<ExecutionPlan, Operation> factory =
                 plan -> new SnapshotOperation(jobId, executionId, newSnapshotId, isTerminal);
 
-        invoke(factory, responses -> onSnapshotCompleted(responses, executionId, newSnapshotId, isTerminal), null);
+        invokeOnParticipants(factory, responses -> onSnapshotCompleted(responses, executionId, newSnapshotId, isTerminal), null);
     }
 
     private void onSnapshotCompleted(Map<MemberInfo, Object> responses, long executionId, long snapshotId,
@@ -551,7 +551,7 @@ public class MasterContext {
         // was successful.
         SnapshotOperationResult mergedResult = new SnapshotOperationResult();
         for (Object response : responses.values()) {
-            // the response either SnapshotOperationResult or an exception, see #invoke() method
+            // the response is either SnapshotOperationResult or an exception, see #invokeOnParticipants() method
             if (response instanceof Throwable) {
                 response = new SnapshotOperationResult(0, 0, 0, (Throwable) response);
             }
@@ -681,7 +681,7 @@ public class MasterContext {
         }
 
         Function<ExecutionPlan, Operation> operationCtor = plan -> new CompleteExecutionOperation(executionId, finalError);
-        invoke(operationCtor, responses -> onCompleteExecutionCompleted(error), null);
+        invokeOnParticipants(operationCtor, responses -> onCompleteExecutionCompleted(error), null);
     }
 
     private void onCompleteExecutionCompleted(Throwable error) {
@@ -814,54 +814,50 @@ public class MasterContext {
      * @param completionCallback a consumer that will receive a map of
      *                           responses, one for each member. The value will
      *                           be either the response or an exception thrown
-     *                           from the operation.
+     *                           from the operation
+     * @param callback A callback that will be added to each individual
+     *                operation for each member
      */
-    private void invoke(Function<ExecutionPlan, Operation> operationCtor,
-                        Consumer<Map<MemberInfo, Object>> completionCallback,
-                        ExecutionCallback<Object> callback) {
-        CompletableFuture<Void> doneFuture = new CompletableFuture<>();
-        Map<MemberInfo, InternalCompletableFuture<Object>> futures = new ConcurrentHashMap<>();
-        invokeOnParticipants(futures, doneFuture, operationCtor);
-
-        // once all invocations return, notify the completion callback
-        doneFuture.whenComplete(withTryCatch(logger, (aVoid, throwable) -> {
-            Map<MemberInfo, Object> responses = new HashMap<>();
-            for (Entry<MemberInfo, InternalCompletableFuture<Object>> entry : futures.entrySet()) {
-                Object val;
-                try {
-                    val = entry.getValue().get();
-                } catch (InterruptedException e) {
-                    Thread.currentThread().interrupt();
-                    val = e;
-                } catch (Exception e) {
-                    val = peel(e);
-                }
-                responses.put(entry.getKey(), val);
-            }
-            completionCallback.accept(responses);
-        }));
-
-        if (callback != null) {
-            futures.values().forEach(f -> f.andThen(callback));
+    private void invokeOnParticipants(Function<ExecutionPlan, Operation> operationCtor,
+                                      Consumer<Map<MemberInfo, Object>> completionCallback,
+                                      ExecutionCallback<Object> callback) {
+        Map<MemberInfo, AtomicReference<Object>> responses = new HashMap<>();
+        // pre-populate the map so that it's read-only from here on and can be used from multiple threads
+        for (MemberInfo member : executionPlanMap.keySet()) {
+            responses.put(member, new AtomicReference<>());
         }
-    }
 
-    private void invokeOnParticipants(Map<MemberInfo, InternalCompletableFuture<Object>> futures,
-                                      CompletableFuture<Void> doneFuture,
-                                      Function<ExecutionPlan, Operation> opCtor) {
-        AtomicInteger remainingCount = new AtomicInteger(executionPlanMap.size());
-        for (Entry<MemberInfo, ExecutionPlan> e : executionPlanMap.entrySet()) {
+        // we'll expect two callbacks for each member
+        AtomicInteger remainingCount = new AtomicInteger(2 * executionPlanMap.size());
+
+        Runnable handleDone = () -> {
+            if (remainingCount.decrementAndGet() == 0) {
+                Map<MemberInfo, Object> responses2 = new HashMap<>();
+                // unwrap the AtomicReferences. Can't use stream api because toMap doesn't support null values...
+                for (Entry<MemberInfo, AtomicReference<Object>> entry : responses.entrySet()) {
+                    responses2.put(entry.getKey(), entry.getValue().get());
+                }
+                completionCallback.accept(responses2);
+            }
+        };
+
+        for (Entry<MemberInfo, AtomicReference<Object>> e : responses.entrySet()) {
             MemberInfo member = e.getKey();
-            Operation op = opCtor.apply(e.getValue());
-            InternalCompletableFuture<Object> future = nodeEngine.getOperationService()
-                 .createInvocationBuilder(JetService.SERVICE_NAME, op, member.getAddress())
-                 .setDoneCallback(() -> {
-                     if (remainingCount.decrementAndGet() == 0) {
-                         doneFuture.complete(null);
-                     }
-                 })
-                 .invoke();
-            futures.put(member, future);
+            AtomicReference<Object> response = e.getValue();
+            Operation op = operationCtor.apply(executionPlanMap.get(member));
+            InternalCompletableFuture<Object> future =
+                    nodeEngine.getOperationService()
+                              .createInvocationBuilder(JetService.SERVICE_NAME, op, member.getAddress())
+                              .setDoneCallback(handleDone)
+                              .setExecutionCallback(callbackOf((r, throwable) -> {
+                                  boolean success = response.compareAndSet(null, r == null ? throwable : r);
+                                  assert success : "response=" + response.get();
+                                  handleDone.run();
+                              }))
+                              .invoke();
+            if (callback != null) {
+                future.andThen(callback);
+            }
         }
     }
 

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/MasterContext.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/MasterContext.java
@@ -54,7 +54,6 @@ import com.hazelcast.spi.impl.NodeEngineImpl;
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
 import java.util.Collection;
-import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
@@ -62,9 +61,10 @@ import java.util.Map.Entry;
 import java.util.Set;
 import java.util.concurrent.CancellationException;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
-import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Consumer;
 import java.util.function.Function;
 import java.util.stream.Collectors;
@@ -89,6 +89,7 @@ import static com.hazelcast.jet.impl.execution.SnapshotContext.NO_SNAPSHOT;
 import static com.hazelcast.jet.impl.execution.init.CustomClassLoadedObject.deserializeWithCustomClassLoader;
 import static com.hazelcast.jet.impl.execution.init.ExecutionPlanBuilder.createExecutionPlans;
 import static com.hazelcast.jet.impl.util.ExceptionUtil.isRestartableException;
+import static com.hazelcast.jet.impl.util.ExceptionUtil.peel;
 import static com.hazelcast.jet.impl.util.ExceptionUtil.withTryCatch;
 import static com.hazelcast.jet.impl.util.Util.callbackOf;
 import static com.hazelcast.jet.impl.util.Util.jobNameAndExecutionId;
@@ -102,6 +103,13 @@ import static java.util.stream.Collectors.toList;
  * shared between multiple executions.
  */
 public class MasterContext {
+
+    private static final Object NULL_OBJECT = new Object() {
+        @Override
+        public String toString() {
+            return "NULL_OBJECT";
+        }
+    };
 
     public static final int SNAPSHOT_RESTORE_EDGE_PRIORITY = Integer.MIN_VALUE;
     public static final String SNAPSHOT_VERTEX_PREFIX = "__snapshot_";
@@ -506,7 +514,7 @@ public class MasterContext {
     private void cancelExecutionInvocations(long jobId, long executionId, TerminationMode mode) {
         nodeEngine.getExecutionService().execute(ExecutionService.ASYNC_EXECUTOR, () ->
                 invokeOnParticipants(plan -> new TerminateExecutionOperation(jobId, executionId, mode),
-                        responses -> { }, null));
+                        null, null));
     }
 
     void beginSnapshot(long executionId) {
@@ -814,42 +822,34 @@ public class MasterContext {
 
     /**
      * @param completionCallback a consumer that will receive a map of
-     *                           responses, one for each member. The value will
-     *                           be either the response or an exception thrown
-     *                           from the operation
-     * @param callback A callback that will be added to each individual
-     *                operation for each member
+     *                           responses, one for each member, after all have
+     *                           been received. The value will be either the
+     *                           response or an exception thrown from the
+     *                           operation
+     * @param callback A callback that will be called after each individual
+     *                operation for each member completes
      */
     private void invokeOnParticipants(Function<ExecutionPlan, Operation> operationCtor,
-                                      Consumer<Map<MemberInfo, Object>> completionCallback,
-                                      ExecutionCallback<Object> callback) {
-        Map<MemberInfo, AtomicReference<Object>> responses = new HashMap<>();
-        // pre-populate the map so that it's read-only from here on and can be used from multiple threads
-        for (MemberInfo member : executionPlanMap.keySet()) {
-            responses.put(member, new AtomicReference<>());
-        }
+                                      @Nullable Consumer<Map<MemberInfo, Object>> completionCallback,
+                                      @Nullable ExecutionCallback<Object> callback) {
+        ConcurrentMap<MemberInfo, Object> responses = new ConcurrentHashMap<>();
         AtomicInteger remainingCount = new AtomicInteger(executionPlanMap.size());
-        for (Entry<MemberInfo, AtomicReference<Object>> e : responses.entrySet()) {
-            MemberInfo member = e.getKey();
-            AtomicReference<Object> response = e.getValue();
-            Operation op = operationCtor.apply(executionPlanMap.get(member));
-            InternalCompletableFuture<Object> future =
-                    nodeEngine.getOperationService()
-                              .createInvocationBuilder(JetService.SERVICE_NAME, op, member.getAddress())
-                              .setExecutionCallback(callbackOf((r, throwable) -> {
-                                  boolean success = response.compareAndSet(null, r == null ? throwable : r);
-                                  assert success : "response=" + response.get();
-                                  if (remainingCount.decrementAndGet() == 0) {
-                                      // unwrap the AtomicReferences. Can't use stream api, toMap doesn't allow null values
-                                      Map<MemberInfo, Object> responses2 = new HashMap<>();
-                                      for (Entry<MemberInfo, AtomicReference<Object>> entry : responses.entrySet()) {
-                                          responses2.put(entry.getKey(), entry.getValue().get());
-                                      }
-                                      completionCallback.accept(responses2);
-                                  }
+        for (Entry<MemberInfo, ExecutionPlan> entry : executionPlanMap.entrySet()) {
+            MemberInfo member = entry.getKey();
+            Operation op = operationCtor.apply(entry.getValue());
+            InternalCompletableFuture<Object> future = nodeEngine.getOperationService()
+                    .createInvocationBuilder(JetService.SERVICE_NAME, op, member.getAddress())
+                    .invoke();
 
-                              }))
-                              .invoke();
+            if (completionCallback != null) {
+                future.andThen(callbackOf((r, throwable) -> {
+                    responses.put(member, r != null ? r : throwable != null ? peel(throwable) : NULL_OBJECT);
+                    if (remainingCount.decrementAndGet() == 0) {
+                        completionCallback.accept(responses);
+                    }
+                }));
+            }
+
             if (callback != null) {
                 future.andThen(callback);
             }

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/operation/SnapshotOperation.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/operation/SnapshotOperation.java
@@ -143,8 +143,8 @@ public class SnapshotOperation extends AsyncJobOperation {
 
         /**
          * Merge other SnapshotOperationResult into this one. It adds the
-         * totals and if the other result has error, it will take it to this,
-         * unless there already was one.
+         * subtotals and if the other result has an error, it will store it
+         * into this, unless this result already has one.
          */
         public void merge(SnapshotOperationResult other) {
             numBytes += other.numBytes;

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/Util.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/Util.java
@@ -24,8 +24,8 @@ import com.hazelcast.jet.JetException;
 import com.hazelcast.jet.JetInstance;
 import com.hazelcast.jet.core.Watermark;
 import com.hazelcast.jet.function.DistributedBiFunction;
-import com.hazelcast.jet.impl.JetService;
 import com.hazelcast.jet.impl.JetEvent;
+import com.hazelcast.jet.impl.JetService;
 import com.hazelcast.logging.ILogger;
 import com.hazelcast.map.AbstractEntryProcessor;
 import com.hazelcast.map.EntryProcessor;
@@ -64,6 +64,7 @@ import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.concurrent.atomic.AtomicLongArray;
+import java.util.function.BiConsumer;
 import java.util.function.Consumer;
 import java.util.function.Supplier;
 import java.util.function.ToIntFunction;
@@ -134,6 +135,32 @@ public final class Util {
             @Override
             public void onFailure(Throwable throwable) {
                 onError.accept(throwable);
+            }
+        };
+    }
+
+    /**
+     * This method will generate an {@link ExecutionCallback} which allows to
+     * asynchronously get notified when the execution is completed, either
+     * successfully or with an error.
+     *
+     * @param callback BiConsumer to call when execution is completed. Only one
+     *                of the passed values will be non-null, except for the
+     *                case the normal result is null, in which case both values
+     *                will be null
+     * @param <T> type of the response
+     * @return {@link ExecutionCallback}
+     */
+    public static <T> ExecutionCallback<T> callbackOf(BiConsumer<T, Throwable> callback) {
+        return new ExecutionCallback<T>() {
+            @Override
+            public void onResponse(T o) {
+                callback.accept(o, null);
+            }
+
+            @Override
+            public void onFailure(Throwable throwable) {
+                callback.accept(null, throwable);
             }
         };
     }

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/core/ManualRestartTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/core/ManualRestartTest.java
@@ -1,4 +1,4 @@
-    /*
+/*
  * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
@@ -18,8 +18,6 @@ package com.hazelcast.jet.core;
 
 import com.hazelcast.client.map.helpers.AMapStore;
 import com.hazelcast.config.MapConfig;
-import com.hazelcast.config.PartitioningStrategyConfig;
-import com.hazelcast.core.PartitioningStrategy;
 import com.hazelcast.jet.JetInstance;
 import com.hazelcast.jet.Job;
 import com.hazelcast.jet.config.JetConfig;

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/core/ManualRestartTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/core/ManualRestartTest.java
@@ -1,4 +1,4 @@
-/*
+    /*
  * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
@@ -18,6 +18,8 @@ package com.hazelcast.jet.core;
 
 import com.hazelcast.client.map.helpers.AMapStore;
 import com.hazelcast.config.MapConfig;
+import com.hazelcast.config.PartitioningStrategyConfig;
+import com.hazelcast.core.PartitioningStrategy;
 import com.hazelcast.jet.JetInstance;
 import com.hazelcast.jet.Job;
 import com.hazelcast.jet.config.JetConfig;
@@ -181,14 +183,14 @@ public class ManualRestartTest extends JetTestSupport {
         job.join();
 
         Map<Integer, Integer> actual = new ArrayList<>(instances[0].<Entry<Integer, Integer>>getList("sink")).stream()
-                .filter(e -> e.getKey() == 0)
+                .filter(e -> e.getKey() == 0) // we'll only check partition 0
                 .map(Entry::getValue)
                 .collect(Collectors.toMap(e -> e, e -> 1, (o, n) -> o + n, TreeMap::new));
 
-        assertEquals(actual.toString(), (Integer) 1, actual.get(0));
-        assertEquals(actual.toString(), (Integer) 1, actual.get(9999));
-        // the result should be some ones, then some twos and then some ones. The ones should be from the time
-        // when the source was replayed from last snapshot.
+        assertEquals("first item != 1, " + actual.toString(), (Integer) 1, actual.get(0));
+        assertEquals("last item != 1, " + actual.toString(), (Integer) 1, actual.get(9999));
+        // the result should be some ones, then some twos and then some ones. The twos should be during the time
+        // since the last successful snapshot until the actual termination, when there was reprocessing.
         boolean sawTwo = false;
         boolean sawOneAgain = false;
         for (Integer v : actual.values()) {

